shell/platform/linux/fl_clipping_view.cc

```diff
@@ -0,0 +1,181 @@
+                         rect->right - bottom_right_radii->width, rect->bottom,
+                         rect->right - bottom_right_radii->width, rect->bottom);
+
+          cairo_line_to(cr, rect->left + bottom_left_radii->width,
+                        rect->bottom);
+          cairo_curve_to(cr, rect->left, rect->bottom,  //
+                         rect->left, rect->bottom - bottom_left_radii->height,
+                         rect->left, rect->bottom - bottom_left_radii->height);
+
+          cairo_line_to(cr, rect->left, rect->top + top_left_radii->height);
+          cairo_curve_to(cr, rect->left, rect->top,
+                         rect->left + top_left_radii->width, rect->top,
+                         rect->left + top_left_radii->width, rect->top);
+
+          cairo_close_path(cr);
+          cairo_clip(cr);
+        } break;
+        case kFlutterPlatformViewMutationTypeTransformation: {
+          cairo_matrix_t matrix = {
+              .xx = mutation->transformation.scaleX,
+              .yx = mutation->transformation.skewY,
+              .xy = mutation->transformation.skewX,
+              .yy = mutation->transformation.scaleY,
+              .x0 = mutation->transformation.transX,
+              .y0 = mutation->transformation.transY,
+          };
+          cairo_transform(cr, &matrix);
+        } break;
+      }
+    }
+  }
+  cairo_translate(cr, -self->geometry.x, -self->geometry.y);
+  gboolean result =
+      GTK_WIDGET_CLASS(fl_clipping_view_parent_class)->draw(widget, cr);
+  cairo_restore(cr);
+  return result;
+}
+
+static void fl_clipping_view_constructed(GObject* object) {
+  gtk_event_box_set_visible_window(GTK_EVENT_BOX(object), TRUE);
+}
+
+static void fl_clipping_view_class_init(FlClippingViewClass* klass) {
+  GObjectClass* gobject_class = G_OBJECT_CLASS(klass);
+  gobject_class->constructed = fl_clipping_view_constructed;
+  gobject_class->dispose = fl_clipping_view_dispose;
+
+  GtkWidgetClass* widget_class = GTK_WIDGET_CLASS(klass);
+  widget_class->draw = fl_clipping_view_draw;
+  widget_class->size_allocate = fl_clipping_view_size_allocate;
+}
+
+static void fl_clipping_view_init(FlClippingView* self) {}
+
+GtkWidget* fl_clipping_view_new() {
+  return GTK_WIDGET(g_object_new(fl_clipping_view_get_type(), nullptr));
+}
+
+void fl_clipping_view_reset(FlClippingView* self,
+                            GtkWidget* child,
+                            GdkRectangle* geometry,
+                            GPtrArray* mutations) {
+  g_return_if_fail(FL_IS_CLIPPING_VIEW(self));
+
+  GtkWidget* old_child = gtk_bin_get_child(GTK_BIN(self));
+  if (old_child != child) {
+    if (old_child) {
+      gtk_container_remove(GTK_CONTAINER(self), old_child);
+    }
+
+    g_object_ref(child);
+    gtk_container_add(GTK_CONTAINER(self), child);
+  }
+  self->geometry = *geometry;
+  g_clear_pointer(&self->mutations, g_ptr_array_unref);
+  self->mutations = g_ptr_array_ref(mutations);
```

Comment:
This will fail if mutations is NULL - the ref should be moved inside the branch below.